### PR TITLE
Do not give up resolving a file if any errors were found in any file

### LINKED ISSUE #830: We should not give up resolving imports
Solang gives up resolving imports if one of them fails. When we build [this contract](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC721/ERC721.sol) from OpenZepellin, Solang fails on `Address.sol` and does not even parse `Context.sol`. This creates inconsistent error messages when we try to resolve the main file.

### PR DESCRIPTION
**Pull request by seanyoung**
Currently, if there are any errors found in any file, then we do not
resolve anything beyond the types.

This is a step towards solang not giving up early when there are errors in the file; we should always continue and resolve as much as possible.

Fixes #830

### CODE REVIEW

File: src/sema/types.rs [hunk 1]
```diff
@@ -220,24 +219,26 @@ pub fn resolve_fields(delay: ResolveFields, file_no: usize, ns: &mut Namespace)
             let def = ns.structs[struct_no].clone();
```
Comment (LucasSte) [inline src/sema/types.rs:219]: Inlining function definitions makes the code unreadable.

File: src/sema/types.rs [hunk 2]
```diff
@@ -247,21 +248,17 @@ pub fn resolve_fields(delay: ResolveFields, file_no: usize, ns: &mut Namespace)
         check(struct_no, &mut vec![struct_no], ns);
     }
 
-    // Do not attempt to call struct offsets if there are any infinitely recursive structs
-    if !ns.diagnostics.any_errors() {
-        struct_offsets(ns);
-    }
+    // Calculate the offset of each field in all the struct types
+    struct_offsets(ns);
```
Comment (LucasSte) [inline src/sema/types.rs:252]: If a struct has itself as a member, solang will enter an infinite loop when calculating offsets. There should a way to stop the loop.

Take this contract as an example:

```solidity
contract tt {
    struct rr {
        uint a;
        rr b;
    }

    rr s_rr;
    function getThis() public returns (uint) {
        return s_rr.a;
    }
}
```

**Review by LucasSte — CHANGES_REQUESTED** [review]

File: src/sema/types.rs (hunk 2 above)
Comment (seanyoung) [reply src/sema/types.rs:252]: Right. Looks like this goes into an infinite loop when compiling for solana, which was not covered by the tests yet. Fixed.

File: src/sema/types.rs (hunk 1 above)
Comment (LucasSte) [reply src/sema/types.rs:219]: https://github.com/hyperledger-labs/solang/blob/779b4b4aab4d307943e7ac95452800758f8eb155/src/sema/types.rs#L219-L245

I think we can inline definition or use lambda functions when they're small sized pieces of code. Here, you defined a large function inside another one.

Comment (seanyoung) [reply src/sema/types.rs:219]: Right, I was totally confused because you are referring to some existing code.

Comment (seanyoung) [reply src/sema/types.rs]: You are right of course, that is not the most readable code. I've moved the function, looks better now (I think)

File: src/sema/types.rs (hunk 2 above)
Comment (LucasSte) [reply src/sema/types.rs:252]: I couldn't find any test case for the recursive struct. I believe it is a good idea to have it.

Comment (seanyoung) [reply src/sema/types.rs:252]: Just look for " has infinite size". There are a few tests in the `tests/contract_tests/`. I've moved one substrate test to solana in this PR.

Comment (seanyoung) [reply src/sema/types.rs:252]: As I said above, this PR moves   substrate test to solana

**Review by LucasSte — APPROVED** [review]

**Pull request merged**